### ISSUE TITLE
Improve API doc for `env` parameter to IReactorProcess.spawnProcess

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3480]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3480|
|Type|enhancement|
|Created|2008-10-18 03:09:40Z|

It's hard to tell what exactly different values for this parameter mean.  As a bonus, it differs on POSIX and Windows.  This should be documented clearly.

Attachments:

* [interfaces.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4ea/4ea7e03e6bbde62f541c82425735ab2d8b013a11/5e969820e4ce0a5287f06b24960f4f6a4dafe305.patch) (1037 bytes) - added by cary on 2009-04-02 16:26:41Z - 
* [IReactorProcess-env.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4ea/4ea7e03e6bbde62f541c82425735ab2d8b013a11/b1e124a96de3dc889aa2613f2d85cd8d58c96154.patch) (1919 bytes) - added by jesstess on 2009-09-23 03:45:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3480 3480
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1224299380000000 1224299380000000
changetime__1260757304000000 1260757304000000
version__None None
owner__ 
cc__jesstess cc__ivank cc__glyph cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="cary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|cary set owner to @exarkun|
|-|-|

exarkun mentioned that some unicode related caveats should be mentioned as well. I'm not savvy on this topic, but did find the following bug:
http://bugs.python.org/issue4006

Please advise.

**Comment by twisted-trac** [comment]
|<img alt="cary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|cary set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="cary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|cary removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to cary|
|-|-|

The patch is unfortunately invalid epytext.  You should install [pydoctor](http://codespeak.net/~mwh/pydoctor/) and use "`pydoctor --auto`" in the root of your Twisted checkout to make sure that docstrings format properly when you're editing them.

It would also nice if it were _super_ explicit about the difference between `None` and not being specified.  For example, "If explicitly passed as None, the child process inherit the environment of this process, via `os.environ`.  If not passed, the behavior varies between platforms."

Thanks for giving these doc tickets some love!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess removed owner|
|-|-|

The new patch is valid with pydoctor and adds glyph's super explicit description.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @jesstess|
|-|-|

Unfortunately still not quite right.  Sorry for the incorrect guidance I gave you: as you described it, the behavior varies between platforms even when the value _is_ passed, and it should say so at the beginning.  As a result, it's not possible to pass an empty environment, or even remove values from the environment, on Windows.  This is really a bug; specifically, bug [#7361](https://github.com/twisted/twisted/issues/7361).  It might even be worthwhile to note that this behavior will be corrected in some future version, although that correction probably won't be called `spawnProcess` for compatibility reasons.

As for the unicode issues mentioned above, it should be clear that env is a dictionary mapping `str` to `str`, since "string" is slightly ambiguous and might mean unicode.  I don't think the unicode issues cary mentioned above need to be specifically mentioned as long as that is clear.

Since I'm delaying your ticket for yet another review cycle, I'll try to provide some more useful information :).  You might also want to consider using some other epytext markup.  You can [use a list to enumerate the different platform behavior](http://epydoc.sourceforge.net/epytext.html#list), for example, or [the @type field to separate out the description of env's type from its purpose](http://epydoc.sourceforge.net/fields.html#fields).  You could also use [a hyperlink](http://epydoc.sourceforge.net/epytext.html#urls) if you want to point at the ticket to correct the divergent behavior between platforms.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|ivank commented|
|-|-|

Windows can have unicode environments (for both keys and values, I believe). See [#3964](https://github.com/twisted/twisted/issues/3964)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess commented|
|-|-|

glyph, I'm a little confused by your latest feedback. Here's a list of the possible scenarios as I understand them. Can you correct me on this? Am I missing any cases?

env = not set:
 1. Windows: pass os.environ
 1. POSIX: pass empty environment
env = dict:
 1. Windows: update os.environ with key/value pairs in dictionary. However, if there are keys with empty values, those environment variables _won't_ be unset
 1. POSIX: update os.environ with the key/value pairs in dictionary
env = None
 1. Windows: pass os.environ
 1. POSIX: pass os.environ

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

jesstess,

You're almost right, except for 'env = dict', point 2.  On a POSIX system, if a dictionary is passed as the environment argument to `reactor.spawnProcess`, the spawned process will have the key/value pairs present in that dictionary as its _complete_ environment.  There is no interaction with `os.environ`.  See [here, for the POSIX implementation](http://twistedmatrix.com/trac/browser/trunk/twisted/internet/process.py?rev=26050#L442) and [here, for the win32 implementation](http://twistedmatrix.com/trac/browser/trunk/twisted/internet/_dumbwin32proc.py?rev=27135#L163) of the relevant propagation.

The language you use in 'env = dict', point 1 suggests that you might also be missing a subtle distinction between the idea of an empty environment variable and an un-set environment variable.

Hopefully this shell session will be illuminating:
```
$ python -c 'import os; print repr(os.environ["FOO"])'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python2.5/UserDict.py", line 22, in __getitem__
    raise KeyError(key)
KeyError: 'FOO'
$ FOO= python -c 'import os; print repr(os.environ["FOO"])'
''
```

Is everything clear now?  If you find some of this helpful, it might be good to copy it to copy it to [#7361](https://github.com/twisted/twisted/issues/7361) to increase the chances that someone else will understand and fix that bug :).

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_12"></a>|@jesstess removed owner|
|-|-|

New patch listing all the behaviors and linking to [#7361](https://github.com/twisted/twisted/issues/7361). I'll add a comment on that ticket about updating the spawnProcess documentation once it gets resolved.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra set owner to @jesstess|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_14"></a>|@jesstess set status to `closed`|
|-|-|

(In [27770]) Apply IReactorProcess-env.patch

Author: jesstess
Reviewers: glyph, ivank, thijs
Fixes: [#3480](https://github.com/twisted/twisted/issues/3480)

Clarify documentation on the env parameter to spawnProcess, in
particular enumerating differences in behavior between POSIX and
Windows systems.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
